server/app/utils.py

```diff
@@ -264,15 +283,128 @@ def assign_submission(backup_id, submit):
         subm.put()
         subm.mark_as_final()
 
-def parse_date(date):
-    # TODO Describe what date translation is happening here. Probably needs
-    # a rewrite to handle daylight savings and work with other time zones.
-    try:
-        date = datetime.datetime.strptime(
-            date, app.config["GAE_DATETIME_FORMAT"])
-    except ValueError:
-        date = datetime.datetime.strptime(
-            date, "%Y-%m-%d %H:%M:%S")
+def sort_by_assignment(key_func, entries):
+    entries = sorted(entries, key=key_func)
+    return itertools.groupby(entries, key_func)
 
-    delta = datetime.timedelta(hours=7)
-    return datetime.datetime.combine(date.date(), date.time()) + delta
+@ndb.toplevel
+def merge_user(user_key, dup_user_key):
+    """
+    Merges |dup_user| into |user|.
+    """
+    if isinstance(user_key, ModelProxy.Base):
+        user = user_key
+        user_key = user_key.key
+        get_user = lambda: dup_user
+    else:
+        user = user_key.get_async()
+        def get_user():
+            if isinstance(user, ndb.Future):
+                user = user.get_result()
+            return user
+
+    if isinstance(dup_user_key, ModelProxy.Base):
+        dup_user = dup_user_key
+        get_dup_user = lambda: dup_user
+        dup_user_key = dup_user_key.key
+    else:
+        dup_user = dup_user_key.get_async()
+        def get_dup_user():
+            if isinstance(dup_user, ndb.Future):
+                dup_user = dup_user.get_result()
+            return dup_user
+
+    # Leave all groups
+    G = ModelProxy.Group
+    groups = G.query(ndb.OR(
+        G.member == dup_user_key,
+        G.invited == dup_user_key)).fetch()
+    for group in groups:
+        group.exit(dup_user_key)
+
+    # Re-submit submissions
+    S = ModelProxy.Submission
+    subms = S.query(S.submitter == dup_user_key).fetch()
+    for subm in subms:
+        subm.resubmit(user_key)
+
+    dup_user = get_dup_user()
+    # Change email
+    # Right now, I just append the hash of their key, which should be
+    # something they can't guess... Maybe make it more secure??
```

Comment:
I'm fine with just having their email twice- it doesn't need to be unguessable, right? It should also be possible to retreive these old users just in case.